editor/src/core/tailwind/tailwind.ts

```diff
@@ -0,0 +1,243 @@
+import { includesDependency } from '../../components/editor/npm-dependency/npm-dependency'
+import { propOrNull } from '../shared/object-utils'
+import { memoize } from '../shared/memoize'
+
+const PostCSSPath = '/postcss.config.js'
+const TailwindConfigPath = '/tailwind.config.js'
+
+function hasRequiredDependenciesForTailwind(packageJsonFile: ProjectFile): boolean {
+  const hasTailwindDependency = includesDependency(packageJsonFile, 'tailwindcss')
+  const hasPostCSSDependency = includesDependency(packageJsonFile, 'postcss')
+
+  return hasTailwindDependency && hasPostCSSDependency
+}
+
+function useGetPackageJson(projectContents: ProjectContentTreeRoot): ProjectFile | null {
+  return React.useMemo(() => packageJsonFileFromProjectContents(projectContents), [projectContents])
+}
+
+function useHasRequiredDependenciesForTailwind(projectContents: ProjectContentTreeRoot): boolean {
+  const packageJsonFile = useGetPackageJson(projectContents)
+  return React.useMemo(() => {
+    return packageJsonFile != null && hasRequiredDependenciesForTailwind(packageJsonFile)
+  }, [packageJsonFile])
+}
+
+function postCSSIncludesTailwindPlugin(postCSSFile: ProjectFile, requireFn: RequireFn): boolean {
+  if (isTextFile(postCSSFile)) {
+    try {
+      const requireResult = requireFn('/', PostCSSPath)
+      if (requireResult?.default != null) {
+        return requireResult?.default?.plugins?.tailwindcss != null
+      }
+    } catch (e) {
+      /* Do nothing */
+    }
+  }
+
+  return false
+}
+
+function useGetPostCSSConfigFile(projectContents: ProjectContentTreeRoot): ProjectFile | null {
+  return React.useMemo(() => getContentsTreeFileFromString(projectContents, PostCSSPath), [
+    projectContents,
+  ])
+}
+
+function usePostCSSIncludesTailwindPlugin(
+  projectContents: ProjectContentTreeRoot,
+  requireFn: RequireFn,
+): boolean {
+  const postCSSFile = useGetPostCSSConfigFile(projectContents)
+  return React.useMemo(() => {
+    return postCSSFile != null && postCSSIncludesTailwindPlugin(postCSSFile, requireFn)
+  }, [postCSSFile, requireFn])
+}
+
+const PreflightKey = 'preflight'
+
+function enablesPreflight(tailwindConfig: any): boolean {
+  const corePlugins = propOrNull('corePlugins', tailwindConfig)
+  if (corePlugins != null && typeof corePlugins === 'object') {
+    if (Array.isArray(corePlugins)) {
+      // This means we have an explicit list of all allowed core plugins
+      return corePlugins.includes(PreflightKey)
+    } else {
+      // We have an object that enables or disables specific plugins - all unlisted plugins are enabled
+      const explicitDisabled = corePlugins[PreflightKey] === false // Tailwind doesn't use truthiness here
+      return !explicitDisabled
+    }
+  }
+
+  return true
+}
+
+function convertTailwindToTwindConfig(tailwindConfig: any): Configuration {
+  const preflightEnabled = enablesPreflight(tailwindConfig)
+
+  return {
+    ...tailwindConfig,
+    preflight: preflightEnabled,
```

Comment:
what is the difference between preflight and non-preflight for the canvas?

Reply:
This is just determining whether or not the [preflight](https://tailwindcss.com/docs/preflight) should be included. It doesn't mess up the styling of the canvas.